### PR TITLE
Use the kind reported by [readdir], thereby saving some calls to [stat]

### PR DESCRIPTION
**Pull request by aalekseyev**
Add a C binding to use the file kind as reported by `readdir` so we don't have to make an extra `stat` system call for every directory entry.
In Jenga we observed that directory listings can be sped up significantly by avoiding stat system calls, which can be expensive on certain file systems (ZFS).

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

### CODE REVIEW

File: src/dune_filesystem_stubs/readdir.c
Comment (nojb) [inline src/dune_filesystem_stubs/readdir.c:59]: Note that `caml_modify` assumes that its first argument is a well-formed block. However, `caml_alloc_small` returns a block filled with garbage. The right way to do it is:
```suggestion
  Field(v_tuple, 0) = v_filename;
  Field(v_tuple, 1) = val_file_type(e->d_type);
```
see Rule 5 at https://caml.inria.fr/pub/docs/manual-ocaml/intfc.html#ss:c-low-level-gc-harmony

Comment (aalekseyev) [reply src/dune_filesystem_stubs/readdir.c:59]: That may explain one of the segfaults that I've seen. Thanks!

**Comment by nojb** [comment]
Just to have an idea, what kind of speedup did you observe?

**Comment by aalekseyev** [comment]
In jenga we're seeing aaround ~8% overall speedup on certain configurations (both for from-clean and for restore-from-cache builds). However, in jenga we're doing some excessive amount of repeated listings because of self-triggering, so in dune we should expect much less.

File: src/dune_filesystem_stubs/readdir.c
Comment (nojb) [inline src/dune_filesystem_stubs/readdir.c:45]: `v_kind` is unused?

Comment (aalekseyev) [reply src/dune_filesystem_stubs/readdir.c]: You're right. Removed it now.

**Comment by aalekseyev** [comment]
We tried this on a build of [Jane Street universe ](https://github.com/janestreet/universe) and we tried a synthetic benchmark that consists mostly of listing files and directories and we saw no noticeable performance improvement.

Looks like the performance improvement we saw on jenga does not translate well to dune, maybe because dune makes its syscalls sequentially and jenga does them in parallel. 

I'm thinking we should not merge this unless it actually improves anything.

**Comment by aalekseyev** [comment]
I poked around a bit more and found that the synthetic benchmark was not done correctly.
In reality, on a directory containing ~100k files this change improves the initial scan of the tree by ~240ms, or by 14% (we get ~1.43 instead of ~1.67).

Taking the rest of the rule computation into account, the win is smaller, but measurable: ~3% win (7.05s instead of 7.3s).

On a small zero build in the universe repo the effect is about 50ms out of 2.6s, so ~5%.

(this is all on ZFS on linux)

@jeremiedimino, given the above, do you think we should merge it, after all?

**Comment by jeremiedimino** [comment]
Yes, it seems worth merging to me.

BTW, a further potential improvement would be to batch the readdir calls on the C side, to do less enter/leave blocking section.

**Comment by aalekseyev** [comment]
Indeed. I tried that too, but it makes the amount of C code significantly larger for a much smaller gain, or at least that's what Jenga benchmarks showed. We'll see how if we need that improvement in the future. 

The Windows-side improvements is another thing we should look at in future.

**Comment by aalekseyev** [comment]
@nojb, you didn't hit "approve", so I'm wondering if you have any reservations about this PR. I remember you said you could look at making the windows changes, but I imagine that's a non-trivial piece of work so we shouldn't block this PR on that?
What do you think?

**Review by nojb — APPROVED** [review]

**Comment by nojb** [comment]
> @nojb, you didn't hit "approve", so I'm wondering if you have any reservations about this PR. I remember you said you could look at making the windows changes, but I imagine that's a non-trivial piece of work so we shouldn't block this PR on that?

No, no reservations, thanks!

Indeed I looked a bit to see if something special could be done under Windows, but it wasn't obvious, and I didn't dig deeper than that.

**Comment by aalekseyev** [comment]
Looks like this PR also fixes #3819 by accident. I added a test, but I'm not adding anything to the changelog because the version with the bug was not yet released.

**Pull request merged**
